File: frontent/screens/visit-screen.js
import React, { useEffect, useState } from "react";
import {
  Alert,
  View,
  Text,
  StyleSheet,
  SafeAreaView,
  FlatList,
  ActivityIndicator,
  TouchableOpacity,
} from "react-native";
import Collapsible from "react-native-collapsible";
import _ from "lodash";

const BASE_URL = "http://192.168.1.11:8000"; // You can change this base URL as needed

const VisitScreen = ({ navigation }) => {
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [expandedId, setExpandedId] = useState(null);
  const [striked, setStriked] = useState({}); // Stan dla przekreślonych linii

  const statusColorMapping = {
    in_progress: "#00ff00", // Green for in progress
    pending: "#ffa500", // Orange for pending
    done: "#ff0000", // Red for done
    default: "#808080", // Default grey color for undefined status
  };

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(`${BASE_URL}/api/visits/`);
        if (!response.ok) {
          throw new Error("Network response was not ok");
        }
        const jsonData = await response.json();
        const sortedData = jsonData.sort(
          (b, a) => new Date(a.date) - new Date(b.date)
        );
        setData(sortedData);
        const strikedData = jsonData.reduce(
          (acc, item) => ({ ...acc, [item.id]: item.striked_lines || {} }),
          {}
        );
        setStriked(strikedData);
        setLoading(false);
      } catch (error) {
        setError(error.message);
        setLoading(false);
      }
    };

    fetchData();
  }, []);

  const confirmStatusChange = (id, currentStatus) => {
    let newStatus;
    switch (currentStatus) {
      case "in_progress":
        newStatus = "pending";
        confirmStatus = "Oczekujący";
        break;
      case "pending":
        newStatus = "done";
        confirmStatus = "Zakończony";
        break;
      case "done":
        newStatus = "in_progress";
        confirmStatus = "W trakcie";
        break;
      default:
        newStatus = "in_progress"; // Set to in_progress if undefined
        break;
    }

    Alert.alert(
      "Potwierdzenie zmiany statusu",
      `Czy na pewno chcesz zmienić status na ${confirmStatus}?`,
      [
        {
          text: "Anuluj",
          style: "cancel",
        },
        { text: "Potwierdź", onPress: () => handleStatusChange(id, newStatus) },
      ]
    );
  };

  const handleStatusChange = async (id, newStatus) => {
    const updatedData = data.map((item) => {
      if (item.id === id) {
        updateStatusInDatabase(id, newStatus); // Send updated status to server
        return { ...item, status: newStatus };
      }
      return item;
    });
    setData(updatedData);
  };

  const updateStatusInDatabase = async (id, newStatus) => {
    try {
      const response = await fetch(`${BASE_URL}/api/visit/${id}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ status: newStatus }),
      });
      if (!response.ok) {
        throw new Error("Failed to update status");
      }
    } catch (error) {
      console.error("Error updating status:", error);
    }
  };

  const toggleStriked = (id, index) => {
    setStriked((prev) => {
      const newStriked = {
        ...prev,
        [id]: {
          ...(prev[id] || {}),
          [index]: !(prev[id] && prev[id][index]),
        },
      };
      updateStrikedOnServer(id, newStriked[id]); // Aktualizuj stan na serwerze
      return newStriked;
    });
  };

  const updateStrikedOnServer = async (id, strikedLines) => {
    try {
      const response = await fetch(
        `${BASE_URL}/api/visit/update-striked/${id}`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ strikedLines }),
        }
      );
      if (!response.ok) {
        const text = await response.text();
        console.error("Server response:", text);
        try {
          const data = JSON.parse(text);
          console.error("Parsed server response:", data);
        } catch (error) {
          console.error("Failed to parse server response:", error);
        }
        throw new Error("Failed to update striked lines");
      }
    } catch (error) {
      console.error("Error updating striked lines:", error);
    }
  };

  const renderItem = ({ item }) => (
    <View style={styles.item}>
      <Text style={styles.title}>
        {item.date} id: {item.id}
      </Text>
      <TouchableOpacity
        style={[
          styles.statusIndicator,
          {
            backgroundColor:
              statusColorMapping[item.status] || statusColorMapping.default,
          },
        ]}
        onPress={() => confirmStatusChange(item.id, item.status)}
      >
        <View style={styles.dot}></View>
      </TouchableOpacity>

      <TouchableOpacity
        onPress={() => setExpandedId(expandedId === item.id ? null : item.id)}
      >
        <View style={styles.header}>
          <Text style={styles.collapsHead}>
            Samochód:{" "}
            {item.cars
              .map(
                (car) =>
                  `${car.brand} ${car.model} ${car.year} \nVIN: ${car.vin}`
              )
              .join(", ")}
          </Text>
        </View>
      </TouchableOpacity>

      <Collapsible collapsed={expandedId !== item.id}>
        <View style={styles.content}>
          <Text style={styles.client}>
            Klient:{" "}
            {item.clients.map(
              (client) => `${client.first_name} ${client.phone}`
            )}
          </Text>
          <Text></Text>
          <Text style={styles.title}>{item.name}</Text>
          {item.description.split(",").map((line, index) => (
            <TouchableOpacity
              key={index}
              onPress={() => toggleStriked(item.id, index)}
            >
              <Text
                style={
                  striked[item.id] && striked[item.id][index]
                    ? styles.striked
                    : undefined
                }
              >
                * {line}
              </Text>
            </TouchableOpacity>
          ))}
        </View>
      </Collapsible>
    </View>
  );

  return (
    <SafeAreaView style={styles.screen}>
      {loading ? (
        <ActivityIndicator size="large" color="#0000ff" />
      ) : error ? (
        <Text style={styles.error}>Failed to load data: {error}</Text>
      ) : (
        <FlatList
          data={data}
          keyExtractor={(item) => item.id.toString()}
          renderItem={renderItem}
        />
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  description: {
    padding: 5,
    fontSize: 18,
    fontWeight: "bold",
  },
  striked: {
    textDecorationLine: "line-through",
  },
  screen: {
    padding: 20,
  },
  item: {
    width: "100%",
    padding: 10,
    paddingRight: 5,
    paddingLeft: 5,
    borderBottomWidth: 1,
    borderBottomColor: "#cccccc",
    flexDirection: "column",
    // alignItems: 'center',
    // justifyContent: 'space-between',
  },
  header: {
    flexDirection: "column",
    alignItems: "left",
    textAlign: "center",
    // justifyContent: 'space-between',
    width: "100%",
    paddingVertical: 10,
    paddingHorizontal: 5,
  },
  content: {
    paddingVertical: 10,
  },
  title: {
    fontWeight: "bold",
    textAlign: "center",
  },
  error: {
    color: "red",
  },
  statusIndicator: {
    width: 24,
    height: 24,
    borderRadius: 12,
    alignSelf: "flex-end",
  },
  dot: {
    width: 24,
    height: 24,
    borderRadius: 12,
  },
  collapsHead: {
    fontSize: 16,
    fontWeight: "bold",
    textTransform: "uppercase",
  },
  client: {
    fontWeight: "bold",
    fontSize: 13,
    textAlign: "center",
  },
});

export default VisitScreen;

File: flutter_frontend/lib/screens/visit_screen.dart
// lib/screens/visit_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../providers/visit_provider.dart';
import 'add_visit_screen.dart';
import '../models/visit.dart';
import 'visit_archive_screen.dart';

class VisitScreen extends StatefulWidget {
  @override
  _VisitScreenState createState() => _VisitScreenState();
}

class _VisitScreenState extends State<VisitScreen> {
  @override
  void initState() {
    super.initState();
    Provider.of<VisitProvider>(context, listen: false).fetchVisits();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Visits'),
        actions: [
          IconButton(
            icon: Icon(Icons.archive),
            onPressed: () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => VisitArchiveScreen()),
              ).then((_) {
                Provider.of<VisitProvider>(context, listen: false).fetchVisits();
              });
            },
          ),
          IconButton(
            icon: Icon(Icons.add),
            onPressed: () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (context) => AddVisitScreen()),
              ).then((_) {
                Provider.of<VisitProvider>(context, listen: false).fetchVisits();
              });
            },
          ),
        ],
      ),
      body: SafeArea(
        child: Consumer<VisitProvider>(
          builder: (context, visitProvider, child) {
            if (visitProvider.loading) {
              return Center(child: CircularProgressIndicator());
            } else if (visitProvider.error != null) {
              return Center(
                  child: Text('Failed to load data: ${visitProvider.error}'));
            } else {
              return ListView.builder(
                itemCount: visitProvider.visits.length,
                itemBuilder: (context, index) {
                  final visit = visitProvider.visits[index];
                  return VisitItem(visit: visit);
                },
              );
            }
          },
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          Navigator.push(
            context,
            MaterialPageRoute(builder: (context) => AddVisitScreen()),
          ).then((_) {
            Provider.of<VisitProvider>(context, listen: false).fetchVisits();
          });
        },
        child: Icon(Icons.add),
      ),
    );
  }
}

class VisitItem extends StatefulWidget {
  final Visit visit;

  VisitItem({required this.visit});

  @override
  _VisitItemState createState() => _VisitItemState();
}

class _VisitItemState extends State<VisitItem> {
  bool _isExpanded = false;

  void _confirmStatusChange(
      BuildContext context, String id, String currentStatus) {
    String newStatus;
    String confirmStatus;
    switch (currentStatus) {
      case "in_progress":
        newStatus = "pending";
        confirmStatus = "Oczekujący";
        break;
      case "pending":
        newStatus = "done";
        confirmStatus = "Zakończony";
        break;
      case "done":
        newStatus = "in_progress";
        confirmStatus = "W trakcie";
        break;
      default:
        newStatus = "in_progress";
        confirmStatus = "W trakcie";
        break;
    }

    showDialog(
      context: context,
      builder: (ctx) => AlertDialog(
        title: Text('Potwierdzenie zmiany statusu'),
        content: Text('Czy na pewno chcesz zmienić status na $confirmStatus?'),
        actions: [
          TextButton(
            child: Text('Anuluj'),
            onPressed: () => Navigator.of(ctx).pop(),
          ),
          TextButton(
            child: Text('Potwierdź'),
            onPressed: () {
              Provider.of<VisitProvider>(context, listen: false)
                  .updateStatus(id, newStatus);
              Navigator.of(ctx).pop();
            },
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final visitProvider = Provider.of<VisitProvider>(context);
    final visit = widget.visit;

    return Card(
      child: Column(
        children: [
          ListTile(
            title: Text(
                '${visit.date} ${visit.cars.map((car) => '${car.brand} ${car.model}\nTablica: ${car.licensePlate}').join(", ")}'),
            trailing: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                GestureDetector(
                  onTap: () =>
                      _confirmStatusChange(context, visit.id, visit.status),
                  child: Container(
                    width: 24,
                    height: 24,
                    decoration: BoxDecoration(
                      color: _statusColorMapping[visit.status] ?? Colors.grey,
                      shape: BoxShape.circle,
                    ),
                  ),
                ),
              ],
            ),
            onTap: () {
              setState(() {
                _isExpanded = !_isExpanded;
              });
            },
          ),
          if (_isExpanded)
            Padding(
              padding:
              const EdgeInsets.symmetric(horizontal: 16.0, vertical: 8.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    'Id: ${visit.id} \nSamochód: ${visit.cars.map((car) => '${car.brand} ${car.model} ${car.year} \nVIN: ${car.vin} \nTablica: ${car.licensePlate}').join(", ")}',
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  SizedBox(height: 8),
                  Text(
                    'Klient: ${visit.cars.map((car) => car.client).map((client) => '${client.firstName} ${client.phone}').join(", ")}',
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  SizedBox(height: 8),
                  Text(
                    'Mechanik: ${visit.mechanics.map((mechanic) => '${mechanic.firstName} ${mechanic.lastName}').join(", ")}',
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  SizedBox(height: 8),
                  Text(
                    visit.name,
                    style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
                  ),
                  SizedBox(height: 8),
                  ...visit.description.split(',').asMap().entries.map((entry) {
                    final index = entry.key;
                    final line = entry.value;
                    final isStriked = visit.strikedLines[index] ?? false;
                    return Row(
                      children: [
                        Checkbox(
                          value: isStriked,
                          onChanged: (bool? value) {
                            final newStrikedLines =
                            Map<int, bool>.from(visit.strikedLines);
                            newStrikedLines[index] = value ?? false;
                            visitProvider.updateStrikedLines(
                                visit.id, newStrikedLines);
                          },
                        ),
                        Expanded(
                          child: Text(
                            line,
                            style: TextStyle(
                              fontSize: 16,
                            ),
                          ),
                        ),
                      ],
                    );
                  }).toList(),
                  ButtonBar(children: [
                    ElevatedButton(
                      onPressed: () {
                        Navigator.of(context).push(
                          MaterialPageRoute(
                            builder: (context) => AddVisitScreen(visit: visit),
                          ),
                        ).then((_) {
                          visitProvider.fetchVisits();
                        });
                      },
                      child: Text('Edytuj'),
                    ),
                    ElevatedButton(
                      onPressed: () {
                        visitProvider.confirmArchiveVisit(context, visit.id, visit.name, visit.description, visit.date, visit.status, visit.cars[0], visit.mechanics[0]);
                      },
                      child: Text('Archiwizuj'),
                    ),
                  ]),
                ],
              ),
            ),
        ],
      ),
    );
  }

  static const Map<String, Color> _statusColorMapping = {
    'in_progress': Colors.green,
    'pending': Colors.orange,
    'done': Colors.red,
    'default': Colors.grey,
  };
}
